<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to Linux - Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    <img src="./images/DRP_Logo_only_square_transparent.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Introduction to Linux</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./00-intro.html"> 
<span class="menu-text">Introduction</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./01-start.html"> 
<span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./02-filesystem.html"> 
<span class="menu-text">Filesystem</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./03-files.html"> 
<span class="menu-text">Files</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./04-processes.html"> 
<span class="menu-text">Processes</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./05-analysis.html" aria-current="page"> 
<span class="menu-text">Analysis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./06-scripts.html"> 
<span class="menu-text">Scripts</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./07-manipulation.html"> 
<span class="menu-text">Exercise</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#bioinformatics-data-formats-and-tools" id="toc-bioinformatics-data-formats-and-tools" class="nav-link active" data-scroll-target="#bioinformatics-data-formats-and-tools">Bioinformatics data formats and tools</a></li>
  <li><a href="#working-with-ngs-data-using-gnu-tools" id="toc-working-with-ngs-data-using-gnu-tools" class="nav-link" data-scroll-target="#working-with-ngs-data-using-gnu-tools">Working with NGS data using GNU tools</a></li>
  <li><a href="#case-study-a-simple-rna-seq-analysis-workflow" id="toc-case-study-a-simple-rna-seq-analysis-workflow" class="nav-link" data-scroll-target="#case-study-a-simple-rna-seq-analysis-workflow">Case study: a simple RNA-Seq analysis workflow</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>



<script src="https://kit.fontawesome.com/ece750edd7.js" crossorigin="anonymous"></script>

<hr>
<p>In this section you will learn how to work with common Next Generation Sequencing (NGS) data formats on the command line.</p>
<section id="bioinformatics-data-formats-and-tools" class="level3">
<h3 class="anchored">Bioinformatics data formats and tools</h3>
<hr>
<div class="key-points">
<h2 class="anchored" data-anchor-id="bioinformatics-data-formats-and-tools">
<i class="fas fa-thumbtack"></i> Key Points
</h2>
<ul>
<li>Many standard formats for storing high throughput sequencing data take the form of structured text files, which are easy to manipulate using standard GNU utilities (built in Unix commands)</li>
<li>Many powerful specialist tools for bioinformatics analysis have been developed to use these formats</li>
</ul>
</div>
<p><br></p>
<p>Many of the most common types of file that you will have to work with as a bioinformatician take the form of structured text files, examples include:</p>
<ul>
<li>Standard formats for representing raw sequences
<ul>
<li><a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a></li>
<li><a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ</a></li>
</ul></li>
<li>Tabular formats for representing aligned reads or genome features
<ul>
<li><a href="https://en.wikipedia.org/wiki/BED_(file_format)">BED</a></li>
<li><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM</a></li>
<li><a href="http://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a></li>
</ul></li>
</ul>
<p>Some other compressed formats, such as <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">BAM</a>, which is a compressed version of the SAM format, can easily be converted to human readable text.</p>
<p>Furthermore, numerous specialist bioinformatics tools have been specifically developed for working with these file formats. For example:</p>
<ul>
<li><a href="https://bedtools.readthedocs.io/">bedtools</a> and <a href="https://bedops.readthedocs.io/en/latest/">bedops</a>, which work with BED files</li>
<li>Various aligners, such as <a href="https://github.com/alexdobin/STAR">STAR</a>, <a href="http://daehwankimlab.github.io/hisat2/">hisat2</a>, and others, which take raw sequences in FASTQ or FASTA format and align them to the genome</li>
<li><a href="http://www.htslib.org/">samtools</a>, which works with SAM files and BAM files
<ul>
<li>BAM files can be viewed in SAM format using the <code>samtools view</code> command</li>
</ul></li>
</ul>
<p><em>Note:</em> These tools are not included in most Linux distributions as standard and typically have to be installed separately.</p>
</section>
<section id="working-with-ngs-data-using-gnu-tools" class="level3">
<h3 class="anchored">Working with NGS data using GNU tools</h3>
<hr>
<div class="key-points">
<h2 class="anchored" data-anchor-id="working-with-ngs-data-using-gnu-tools">
<i class="fas fa-thumbtack"></i> Key Points
</h2>
<ul>
<li>It is possible to perform a wide range of complex analysis tasks on NGS data files using standard GNU utilities (built in Unix commands)</li>
<li>In this section we illustrate the application of these tools to NGS data</li>
</ul>
</div>
<p><br></p>
<p>In previous sections, we extracted an archive named <code>bioinformatics_on_the_command_line_files.tar.gz</code> into the <code>~/course</code> directory, creating a directory called <code>~/course/bioinformatics_on_the_command_line_files</code>.</p>
<p>This directory contains a file called <code>yeast_genes.bed</code>, which lists the genomic co-ordinates of 7,126 yeast genes in BED format, and another file called <code>yeast_genome.fasta</code>, which contains the yeast EF4 genome sequence in FASTA format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core1:~/course$</span> head <span class="at">-5</span> bioinformatics_on_the_command_line_files/yeast_genes.bed</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">chrI</span>    334 649 YAL069W .   +</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">chrI</span>    537 792 YAL068W-A   .   +</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">chrI</span>    1806    2169    YAL068C .   <span class="at">-</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">chrI</span>    2479    2707    YAL067W-A   .   +</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">chrI</span>    7234    9016    YAL067C .   <span class="at">-</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core1:~/course$</span> head <span class="at">-5</span> bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span>I</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACA</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">CATCCTAACACTACCCTAACACAGCCCTAATCTAACCCTGGCCAACCTGTCTCTCAACTT</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">ACCCTCCATTACCCTGCCTCCACTCGTTACCCTGTCCCATTCAACCATACCACTCCGAAC</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">CACCATCCATCCCTCTACTTACTACCACTCACCCACCGTTACCCTCCAATTACCCATATC</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ex">[USERNAME]@bifx-core1:~/course$</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="checking-chromosome-names-in-bed-and-fasta-files" class="level4">
<h4 class="anchored" data-anchor-id="checking-chromosome-names-in-bed-and-fasta-files">Checking chromosome names in BED and FASTA files</h4>
<p>Looking at the above outputs, we can see that the naming convention for the chromosomes in the BED file (shown in the first column) appears to be different to the naming convention for the chromosomes in the FASTA file (shown in the first line after the <code>&gt;</code> character). In order to confirm this we would like to produce a sorted list of chromosome names in each file and compare them.</p>
<p>The example below demonstrates the following new commands:</p>
<ul>
<li><code>cut</code>, which we use to select specified fields (or columns) with the <code>-f</code> option, and specified characters with the <code>-c</code> option</li>
<li><code>sort</code>, which sorts the lines it receives from <strong>STDIN</strong>. The <code>-u</code> flag tells it to remove duplicate lines from the output</li>
<li><code>diff</code>, which compares two text files, and outputs the differences between them</li>
</ul>
<p>We can do this as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Move to the relevant directory</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course/bioinformatics_on_the_command_line_files</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Select the first column of the bed file, sort entries and remove duplicates. Redirect output to a file.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> yeast_genes.bed <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> yeast_genes_bed_chromosomes.list </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Select the sequence ID lines from the fasta file. Sort and remove duplicates and select everything from the second character. Redirect to a file.</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">'^&gt;'</span> yeast_genome.fasta <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-c2-</span> <span class="op">&gt;</span> yeast_genome_fasta_chromosomes.list</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co">## Check if both files have the same number of chromosomes</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> <span class="pp">*</span>_chromosomes.list</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co">## Check for differences in the files</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">diff</span> yeast_genes_bed_chromosomes.list yeast_genome_fasta_chromosomes.list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see that there is a mismatch between the chromosome names in the BED and FASTA files. Each chromosome name in the BED file is equivalent to the corresponding name in the FASTA file with ‘chr’ added to the start.</p>
<p>Before using these files in a bioinformatics analysis, we need to update one of them so that the names match. In the following example we fix the BED file by removing ‘chr’ from the start of each line.</p>
<p>The example below uses the <code>sed</code> command, which is used to perform a search and replace style substitution on each line in the BED file using a regular expression. Here, as in the previous examples using <code>grep</code>, the ‘^’ character is a regular expression character representing the start of the line. We then confirm that the chromosomes are now the same using the <code>diff</code> command. This produces no output, which means that there are no differences between the chromosome lists.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Remove "chr" from the start of every line in the bed file</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">'s/^chr//'</span> yeast_genes.bed <span class="op">&gt;</span> yeast_genes.fixed.bed</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Extract the chromsome names</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> yeast_genes.fixed.bed <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> yeast_genes_fixed_bed_chromosomes.list</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Check for differences</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">diff</span> yeast_genes_fixed_bed_chromosomes.list yeast_genome_fasta_chromosomes.list</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co">## Remove the temporary list files</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-i</span> <span class="pp">*</span>.list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="manipulating-bed-files-with-awk-and-sort" class="level4">
<h4 class="anchored">Manipulating BED files with <code>awk</code> and <code>sort</code></h4>
<p>Many standard formats for representing NGS data take the form of tabular files, in which each line contains a number of fields, separated by a particular character (generally a <em>tab</em> character). The <code>yeast_genes.fixed.bed</code> file generated in the previous example fits this pattern.</p>
<p>This example demonstrates how to use <code>awk</code> and <code>sort</code> to find the name and length of the longest gene on chromosome ‘I’ in the <code>yeast_genes.fixed.bed</code> file:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Use the unix programming language awk for text manipulation</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">awk</span> <span class="at">-F</span><span class="st">'\t'</span> <span class="at">-v</span> OFS=<span class="st">'\t'</span> <span class="st">'$1=="I" {print $4,$3-$2}'</span> yeast_genes.fixed.bed <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k2,2nr</span> <span class="kw">|</span> <span class="fu">head</span> <span class="at">-1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output should be <code>YAR050W 4614</code>.</p>
<p>The above example showcases the power of <code>awk</code> in dealing with tabular data. The command <code>awk -F'\t' -v OFS='\t' '$1=="I" {print $4,$3-$2}'</code> can be deconstructed as follows:</p>
<ul>
<li><code>-F'\t'</code> tells <code>awk</code> that the field separator in the input lines is the <em>tab</em> character (<code>\t</code>)</li>
<li><code>-v OFS='\t'</code> tells <code>awk</code> that the <em>tab</em> character should also be used to separate the fields in the output file</li>
<li><code>'$1=="I" {print $4,$3-$2}'</code> is an <code>awk</code> <em>program</em> consisting of a single line. Each line of an <code>awk</code> program is a pair with the structure <em>condition {action}</em>, and the program is run on each line of the input. In this example:
<ul>
<li>The <em>condition</em> is <code>$1=="I"</code>, which tells <code>awk</code> that the <em>action</em> should be performed if the first field is equal to “I”</li>
<li>The <em>action</em> is <code>print $4,$3-$2</code>, which tells <code>awk</code> to output a line to <strong>STDOUT</strong> in which the first field is the 4th field of the input line, and the second field is the result of subtracting the 2nd field from the 3rd. In this case the result is the gene length</li>
</ul></li>
</ul>
<p><em>Note:</em> A good resource to learn more about <code>awk</code> is <a href="https://www.gnu.org/software/gawk/manual/">Effective AWK Programming</a>.</p>
<p>The <code>sort</code> command in the above example includes the option <code>-k2,2nr</code>. <code>-k</code> tells <code>sort</code> to sort by a key, which comprises a start and stop column number, followed by two options, <code>n</code>, which tells <code>sort</code> that the column contains numbers, and <code>r</code>, which tells <code>sort</code> that the lines should be sorted in reverse (i.e.&nbsp;descending) order.</p>
<div class="challenge">
<h2 class="anchored" data-anchor-id="manipulating-bed-files-with-awk-and-sort">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>You could also remove ‘chr’ from the start of each line of <code>yeast_genes.bed</code> by removing the first three characters of every line. See if you can achieve this with the <code>cut</code> command.</p>
<details>
<summary>
Solution
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>Since we know that ‘chr’ is at the start of every line, <code>cut -c 4- yeast_genes.bed</code> would also work.</p>
</div>
</details>
</div>
<p><br></p>
<div class="challenge">
<h2 class="anchored">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>How would you find the shortest gene on the plus strand of chromosome ‘II’ in <code>yeast_genes.fixed.bed</code> using <code>awk</code>?</p>
<details>
<summary>
Solution
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>You could run <code>awk -F'\t' -v OFS='\t' '$1=="II"&amp;&amp;$6=="+" {print $4,$3-$2}' yeast_genes.fixed.bed | sort -k2,2n | head -1</code></p>
</div>
</details>
</div>
</section>
</section>
<section id="case-study-a-simple-rna-seq-analysis-workflow" class="level3">
<h3 class="anchored">Case study: a simple RNA-Seq analysis workflow</h3>
<hr>
<div class="key-points">
<h2 class="anchored" data-anchor-id="case-study-a-simple-rna-seq-analysis-workflow">
<i class="fas fa-thumbtack"></i> Key Points
</h2>
<ul>
<li>It is possible to perform a simple bioinformatics analysis from end to end using only the bash command line</li>
<li>This section presents a simple case study using the <code>STAR</code> aligner and <code>bedtools</code></li>
</ul>
</div>
<p><br></p>
<p>In this section, we will work through a simple pipeline for analysing RNA-Seq data, which involves the following steps:</p>
<ul>
<li>Start with unaligned reads in FASTQ format</li>
<li>Align the reads against an index generated from a target genome, whose sequence is stored in FASTA format, obtaining an output file in BAM format
<ul>
<li>Here we use the yeast EF4 genome, and align the reads using <a href="https://github.com/alexdobin/STAR">STAR</a></li>
</ul></li>
<li>Compute the genome coverage of the aligned reads, obtaining an output file in bedGraph format, which can then be viewed in a genome browser
<ul>
<li>Here we generate the bedGraph file directly from the BED file using <a href="https://bedtools.readthedocs.io/">bedtools</a></li>
</ul></li>
<li>Count the overlaps between the aligned reads and genomic features, stored in BED format, obtaining an output file in BED format, and find the genes with the largest number of overlapping reads
<ul>
<li>Here we use <a href="https://bedtools.readthedocs.io/">bedtools</a> to compute the intersection between the genes and the reference, and use standard GNU utilities to find the genes with the most hits</li>
</ul></li>
</ul>
<p>This analysis uses the files in the ‘bioinformatics_on_the_command_line_files’ directory that we have already been working with, and can be performed as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Move to the course directory</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a new folder called analysis and move into it</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> analysis</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> analysis</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a new folder 00_source_data and move into it</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 00_source_data</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 00_source_data</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co">## Link the yeast fastq file to this folder</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/raw_yeast_rnaseq_data.fastq </span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co">## Move back to the analysis folder and inspect the directory</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a directory for the STAR aligner index files and link the yeast fasta file</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 01_star_index</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 01_star_index</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/yeast_genome.fasta</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co">## Run STAR genomeGenerate to create a mapping index</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--runThreadN</span> 5 <span class="at">--runMode</span> genomeGenerate <span class="at">--genomeDir</span> . <span class="at">--genomeFastaFiles</span> ./yeast_genome.fasta <span class="at">--genomeSAindexNbases</span> 10</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a directory for the alignment outputs</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 02_aligned_reads</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 02_aligned_reads</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="co">## Run STAR to map the reads to the yeast index</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--genomeDir</span> ../01_star_index/ <span class="at">--readFilesIn</span> ../00_source_data/raw_yeast_rnaseq_data.fastq <span class="at">--runThreadN</span> 2 <span class="at">--outFileNamePrefix</span> raw_yeast_rnaseq_data. <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a directory for coverage files</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 03_coverage</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 03_coverage</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="co">## Use bedtools genomecov to convert bam files to bedgraph</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> genomecov <span class="at">-ibam</span> ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam <span class="at">-bg</span> <span class="op">&gt;</span> raw_yeast_rnaseq_data.genomecov.bg</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a><span class="co">## Make a directory for counting reads over genes</span></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> 04_gene_overlap_counts</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> 04_gene_overlap_counts</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a><span class="co">## Link the yeast gene bed file</span></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> ../../bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a><span class="co">## Use bedtools to count intersection of genes and sequencing reads</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-c</span> <span class="at">-a</span> yeast_genes.fixed.bed <span class="at">-b</span> ../02_aligned_reads/raw_yeast_rnaseq_data.Aligned.sortedByCoord.out.bam <span class="kw">|</span> <span class="fu">awk</span> <span class="at">-F</span><span class="st">'\t'</span> <span class="st">'$7&gt;0'</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k7,7nr</span> <span class="op">&gt;</span> raw_yeast_overlap_data.gene_overlap_counts.bed</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a><span class="co">## Output the 10 genes with the highest read counts</span></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-10</span> 04_gene_overlap_counts/raw_yeast_overlap_data.gene_overlap_counts.bed</span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output should look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 460922  466869  RDN37-2 .   <span class="at">-</span>   3730</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 451785  457732  RDN37-1 .   <span class="at">-</span>   3582</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 468812  468931  RDN5-2  .   +   3063</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 468826  468958  YLR154C-H   .   <span class="at">-</span>   3063</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 472464  472583  RDN5-3  .   +   3060</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 472478  472610  YLR156C-A   .   <span class="at">-</span>   3060</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 482044  482163  RDN5-4  .   +   3055</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 482058  482190  YLR157C-C   .   <span class="at">-</span>   3055</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 485696  485815  RDN5-5  .   +   3052</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="ex">XII</span> 485710  485842  YLR159C-A   .   <span class="at">-</span>   3052</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Note:</em> In this analysis we have done everything from scratch, including creating the genome index. For real analyses it is a good idea to use a pre-generated index, as indices for larger genomes take up a lot of space on the server.</p>
<div class="challenge">
<h2 class="anchored">
<i class="fas fa-pencil-alt"></i> Challenge:
</h2>
<p>We saw earlier that the longest gene on chromosome I is YAR050W. Use <code>bedtools getfasta</code> to find the nucleotide sequence of this gene in FASTA format.</p>
<details>
<summary>
Solution
</summary>
<div class="proof solution">
<p><span class="proof-title"><em>Solution</em>. </span></p>
<h2 class="anchored">
<i class="far fa-eye"></i> Solution:
</h2>
<p>Running <code>bedtools getfasta --help</code> shows us that we need to specify an input DNA FASTA file and a BED file with the feature co-ordinates. <br> We can extract the feature co-ordinates of YAR050W from <code>yeast_genes.fixed.bed</code> using <code>grep or awk</code>, and use the output of this command with <code>bedtools getfasta</code>. An example is illustrated in the following example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"YAR050W"</span> bioinformatics_on_the_command_line_files/yeast_genes.fixed.bed <span class="op">&gt;</span> YAR050W.bed</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> getfasta <span class="at">-fi</span> bioinformatics_on_the_command_line_files/yeast_genome.fasta <span class="at">-bed</span> YAR050W.bed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output should look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span>I:203402-208016</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">ATGACAATGCCTCATCGCTATATGTTTTTGGCAGTCTTTACACTTCTGGCACTAACTA...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</details>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>